graph/src/ext/futures.rs

```diff
@@ -97,8 +97,30 @@ pub struct CancelHandle {
     guard: Weak<Mutex<Vec<oneshot::Sender<()>>>>,
 }
 
-impl CancelHandle {
-    pub fn is_canceled(&self) -> bool {
+pub trait CancelToken {
+    fn is_canceled(&self) -> bool;
+    fn check_cancel(&self) -> Result<(), Canceled> {
+        if self.is_canceled() {
+            Err(Canceled)
+        } else {
+            Ok(())
+        }
+    }
+}
+
+pub struct NeverCancel;
+
+impl CancelToken for NeverCancel {
+    #[inline]
+    fn is_canceled(&self) -> bool {
+        false
+    }
+}
+
+pub struct Canceled;
```

Comment:
`CancelableError` is a generic type, so that would cause `check_cancel` to also be generic, but then type inference wouldn't know what type `t.check_cancel()?;` should return - `CancelableError<???>`. This way fixes that issue, but also allows `Canceled` to coerce to other types. For example you could sometimes just have `Result<T, Canceled>` if the operation was otherwise infallible.